src/modules/formBuilder/components/FormBuilder.tsx

```diff
@@ -0,0 +1,224 @@
+  SetStateAction,
+  useCallback,
+  useMemo,
+  useState,
+} from 'react';
+import ConfirmationDialog from '@/components/elements/ConfirmationDialog';
+import Loading from '@/components/elements/Loading';
+import ErrorAlert from '@/modules/errors/components/ErrorAlert';
+import { ErrorState } from '@/modules/errors/util';
+import { updateFormItem } from '@/modules/form/util/formUtil';
+import FormBuilderHeader from '@/modules/formBuilder/components/FormBuilderHeader';
+import FormBuilderPalette from '@/modules/formBuilder/components/FormBuilderPalette';
+import FormItemEditor from '@/modules/formBuilder/components/FormItemEditor';
+import FormTree from '@/modules/formBuilder/components/formTree/FormTree';
+import {
+  DisabledDisplay,
+  EnableBehavior,
+  FormDefinitionFieldsForEditorFragment,
+  FormDefinitionJson,
+  FormItem,
+  UpdateFormDefinitionMutation,
+} from '@/types/gqlTypes';
+
+interface FormBuilderProps {
+  formDefinition: FormDefinitionFieldsForEditorFragment;
+  workingDefinition?: FormDefinitionJson;
+  setWorkingDefinition?: Dispatch<
+    SetStateAction<FormDefinitionJson | undefined>
+  >;
+  errorState?: ErrorState;
+  onSave: (
+    formDefinition: FormDefinitionJson
+  ) => Promise<FetchResult<UpdateFormDefinitionMutation>>;
+  saveLoading: boolean;
+  lastUpdatedDate?: string;
+  lastUpdatedBy?: string;
+  selectedItem?: FormItem;
+  setSelectedItem: Dispatch<SetStateAction<FormItem | undefined>>;
+  originalLinkId?: string;
+  setOriginalLinkId: Dispatch<SetStateAction<string | undefined>>;
+  closeItemEditor: () => void;
+}
+
+const FormBuilder: React.FC<FormBuilderProps> = ({
+  formDefinition,
+  workingDefinition,
+  setWorkingDefinition,
+  errorState,
+  onSave,
+  saveLoading,
+  lastUpdatedDate,
+  lastUpdatedBy,
+  selectedItem,
+  setSelectedItem,
+  originalLinkId,
+  setOriginalLinkId,
+  closeItemEditor,
+}) => {
+  const dirty = useMemo(() => {
+    return !isEqual(workingDefinition, formDefinition.definition);
+  }, [workingDefinition, formDefinition.definition]);
+
+  const [isBlocked, setIsBlocked] = useState(false);
+  const [unblockCallback, setUnblockCallback] = useState<
+    (() => void) | undefined
+  >(undefined);
+
+  const onClickPreview = useCallback(() => {
+    setIsBlocked(dirty);
+    // TODO(#6091) - add another action here and use setUnblockCallback if the form is dirty.
+    //  See example under FormTree's onEditClick below
+  }, [setIsBlocked, dirty]);
+
+  if (!workingDefinition || !setWorkingDefinition) return <Loading />;
+
+  return (
+    <>
+      <ConfirmationDialog
+        open={isBlocked}
+        loading={saveLoading}
+        confirmText='Save changes'
+        cancelText='Continue editing'
+        title='Unsaved changes'
+        onConfirm={() => {
+          onSave(workingDefinition).then(() => {
+            setIsBlocked(false);
+            if (unblockCallback) unblockCallback(); // todo @Martha typescript
+            setUnblockCallback(undefined);
+          });
```

Comment:
Maybe pull this out to an `const onConfirmSave = useCallback(() => ...`. Same for the `onItemClick ` callback. For me I'm not sure how much of this is a style thing vs actually needing the memoization... it's probably fine as-is, but I find it easier to reason about the component when the logic is not dispersed throughout the jsx.